---
title: "Human Genomics"
subtitle: "Session 2: Population Genomics"
author:
  - name: "Palle Duun Rohde"
    url: https://pdrohde.github.io/
    affiliation: Genomic Medicine, Aalborg University
    email: palledr@hst.aau.dk
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/fs61oo/Documents/GitHub/AAU-human-genomics/")
library(kableExtra)
```

### Part I
#### Exercise 1
Population geneticists analyze the extent and patterns of genetic variation within populations, the distribution of genotypes, and the influence of evolutionary forces like natural selection and genetic drift. These factors drive evolutionary change and speciation. Based on your current understanding, how would you address the following questions?

1.  How can we measure genetic variation within a population?
2.  How can we determine whether a population’s genetic structure is changing over time?
<br />
<br />

#### Exercise 2
Which term refers to the complete set of genetic information shared by all individuals within a population?

1.  Gene pool
2.  Genome
3.  Chromosome complement
4.  Breeding unit
5.  Race

Various mechanisms help preserve genetic diversity within a population. Why is maintaining this diversity beneficial?

1.  Homozygosity provides an evolutionary advantage.
2.  Increased diversity promotes inbreeding benefits.
3.  Genetic diversity enhances a population’s ability to adapt to environmental changes.
4.  Greater genetic diversity increases the likelihood of haploidy.
5.  Genetic diversity prevents diploidy.
<br />
<br />

#### Exercise 3
Human serum contains a protein called haptoglobin, which binds to hemoglobin following the lysis of erythrocytes. In some populations, three electrophoretic haptoglobin variants are observed, determined by three autosomal co-dominant alleles: A, B, and C. Among 500 individuals, the following genotype counts were recorded:

<style type="text/css">
.tg  {border-collapse:collapse;border-color:#ccc;border-spacing:0;}
.tg td{background-color:#fff;border-color:#ccc;border-style:solid;border-width:1px;color:#333;
  font-family:Arial, sans-serif;font-size:14px;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{background-color:#f0f0f0;border-color:#ccc;border-style:solid;border-width:1px;color:#333;
  font-family:Arial, sans-serif;font-size:14px;font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-baqh{text-align:center;vertical-align:top}
.tg .tg-amwm{font-weight:bold;text-align:center;vertical-align:top}
.tg .tg-0lax{text-align:left;vertical-align:top}
</style>
<table class="tg"><thead>
  <tr>
    <th class="tg-amwm" colspan="6">No individuals</th>
    <th class="tg-amwm">Total</th>
  </tr></thead>
<tbody>
  <tr>
    <td class="tg-baqh">AA</td>
    <td class="tg-baqh">AB</td>
    <td class="tg-baqh">AC</td>
    <td class="tg-baqh">BB</td>
    <td class="tg-baqh">BC</td>
    <td class="tg-baqh">CC</td>
    <td class="tg-0lax"></td>
  </tr>
  <tr>
    <td class="tg-baqh">109</td>
    <td class="tg-baqh">123</td>
    <td class="tg-baqh">128</td>
    <td class="tg-baqh">33</td>
    <td class="tg-baqh">70</td>
    <td class="tg-baqh">37</td>
    <td class="tg-baqh">500</td>
  </tr>
</tbody>
</table>
<br />

1.  Compute the genotype frequencies within the sample.
2. Compute the allele frequency within the sample.
<br />
<br />

#### Exercise 4
Cystic fibrosis (CF) is an autosomal recessive disease. In individuals who are homozygous for the recessive allele, a malfunction in salt transport occurs in certain cell types, such as those in the lungs. This defect leads to thick mucus buildup, increasing the risk of bacterial lung infections.

Historically, only a small number of individuals with CF survived into adulthood. In Europe, the birth prevalence of CF has been approximately 1 in 2,500 newborns. This frequency is maintained by a balance between mutation and natural selection.

In a European sample, cystic fibrosis (CF) carrier status was analyzed based on three genotypes: 16,890 individuals were homozygous for the normal allele (<tt>RR</tt>), 855 individuals were heterozygous carriers (<tt>Rr</tt>), and 18 individuals were homozygous for the recessive CF allele (<tt>rr</tt>).

1.  Calculate the frequency of each allele in the sample.
<br />
<br />

#### Exercise 5
A geneticist studying a human population discovers that a recessive allele (<tt>s</tt>) associated with sickle cell disease is present in a community of 1,024 individuals. Individuals carrying two copies of the recessive allele (<tt>ss</tt>) develop sickle cell disease, while those with at least one dominant allele (<tt>S</tt>) have normal red blood cells. Within the population, 1 in 256 individuals has sickle cell disease, indicating they are homozygous recessive (<tt>ss</tt>).

1.	Assuming random mating and no selection for this gene, how many individuals with normal red blood cells are carriers of the recessive allele?
<br />
<br />

#### Exercise 6 
This exercise should be solved using <tt>R</tt>.

Run every step in the code below, and explain what each step does. 
```{r, eval=F}
# define alleles
  A1 <- "A"
  A2 <- "a"

# define frequency of A1
  p <- 0.2
  allele <- c(rep(A1,p*10),rep(A2,(1-p)*10))

# define population size
  N <- 100

# define population
  pop <- matrix(nrow=N, ncol=2)
  
  for(i in 1:nrow(pop)){
    pop[i, 1] <- sample(allele, 1)
    pop[i, 2] <- sample(allele, 1)
  }

# combine alleles into genotype
  pop <- cbind(pop, paste(pop[,1],pop[,2],sep=""))
```
<br />

We can count how many we have of each genotype using the `table(snp)` function.

1.  Determine the genotype and allele frequency. 
2.  Make sure you understand every step.

```{r, eval=F}
# genotype count
  obs.count <- table(pop[,3])

# combine the two heterozygote genotypes (Aa and aA is the same)
  pop[,3][pop[,3]=="aA"] <- "Aa"
  obs.count <- table(pop[,3])

# total number of observations
  Nobs <- sum(obs.count)

# genotype frequencies
  geno.freq <- obs.count / Nobs

# allele frequencies
  p.A <- (2*obs.count["AA"] + obs.count["Aa"]) / (2*Nobs)
  p.a <- (2*obs.count["aa"] + obs.count["Aa"]) / (2*Nobs)
```
<br />

3.  Does it match the expectation based on how you simulated the genotypes? 
4.  We can redo the simulations a number of times, and with different population sizes.
```{r, eval=F}
# set global parameters
  N <- seq(10,1000,10)
  rep <- 10
  A1 <- "A"
  A2 <- "a"
  p <- 0.2
  allele <- c(rep(A1,p*10),rep(A2,(1-p)*10))

  p.A <- matrix(nrow=rep, ncol=length(N))  
  ´
# run simulations
  for(i in 1:length(N)){
    for(j in 1:rep){
      pop <- matrix(nrow=N[i], ncol=2)  
      
      for(k in 1:nrow(pop)){
        pop[k, 1] <- sample(allele, 1)
        pop[k, 2] <- sample(allele, 1)
      }
    pop <- cbind(pop, paste(pop[,1],pop[,2],sep=""))
    pop[,3][pop[,3]=="aA"] <- "Aa"
    
    obs.count <- table(pop[,3])
    Nobs <- sum(obs.count)
    geno.freq <- obs.count / Nobs
    p.A[j,i] <- (2*obs.count["AA"] + obs.count["Aa"]) / (2*Nobs)
    }
  }

```

5.  Why does the matrix `p.A` contains many values that are `NA` ?
6.  Visualise the results from these simulations.

<br />
<br />

### Part II












