---
title: "Human Genomics"
subtitle: "Session 7: Polygenic Scores (PGS)"
author:
  - name: "Palle Duun Rohde & Peter Sørensen"
    url: https://pdrohde.github.io/
    affiliation: Genomic Medicine, Aalborg University
    email: palledr@hst.aau.dk
output: html_document
---
<style type="text/css">

body, td {
   font-size: 14px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
---
```{r setup, include=FALSE} 
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/fs61oo/Documents/GitHub/AAU-human-genomics/exercises/PGS/")
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
library(data.table)
library(qgg)
```

The aim of this practical is to provide a simple introduction to computing polygenic scores (PGS) of complex traits and diseases. The practical will be a mix of theoretical and practical exercises in R that are used for illustrating/applying the theory presented in the corresponding lecture on PGS:

* Data used for computing polygenic scores
* Methods  used for computing polygenic scores
* Methods used for evaluating the predictive ability of the polygenic scores

This practical provides a step-by-step guide to performing basic PGS analyses including the following sessions: 

* Session 1: Use R for downloading data
* Session 2: Prepare and explore phenotype data
* Session 3: Prepare and perform quality control of genetic data
* Session 4: Compute GWAS summary statistics 
* Session 5: Compute sparse LD matrices 
* Session 6: Compute PGS using clumping and thresholding (C+T) 
<br />

#### Polygenic scores
Polygenic scores combine information from large numbers of markers across the genome (hundreds to millions) to give a single numerical score for an individual’s relative genetic predisposition for a specific disease or trait on the basis of the DNA variants they have inherited. 

For a particular disease or trait a PGS is calculated as:
			$$PGS=\sum_{i=1}^mX_i  \hat{b}_i$$
where $X_i$ is the genotype vector, and $\hat{b}_i$ the weight of the i’th single genetic marker.

Genomic prediction has been used for many years in animal and plant breeding (e.g., Meuwissen  et al. 2001), and genomic prediction has gained popularity during the last decade because of:

* Larger GWAS sample size = more precision for effect estimates
* Development of methods that combine genome-wide sets of variants
* Large biobanks for validation and testing of genetic risk scores 
* Ability to identify clinically meaningful increases in disease risk predictions 

#### Terminology
Polygenic risk scores, polygenic scores, genomic risk score, genetic scores, genetic predispostion, genetic value,  genomic breeding value is (more or less) the same thing.
<br />

#### Complex traits and diseases
For many complex traits and diseases there will be thousands of genetic variants that each contribute with a small effect on the disease risk or quantitative trait. Rare variant with large effects will only explain small proportion of $h^2$ (low predictive potential). Common variants with small effects can explain larger proportion of $h^2$ (high predictive potential). The majority of complex traits and common diseases in humans are heritable. The heritability determines the value of using genetics for risk prediction. In general, large data sets are required to obtain accurate marker estimates of small to moderate effects, which also improves the prediction accuracy.
<br />

#### Heritability
The heritability ($h^2$) quantify the degree of variation in a phenotypic trait in a population that is due to genetic variation between individuals in that population. It measures how much of the variation of a trait can be attributed to variation of genetic factors, as opposed to variation of environmental factors. The narrow sense heritability is the ratio of additive genetic variance ($\sigma^2_{a}$) to the overall phenotypic variance ($\sigma^2_{y}=\sigma^2_{a}+\sigma^2_{e}$):
\begin{align}
h^2 &= \sigma^2_{a}/(\sigma^2_a+\sigma^2_e)
\end{align}
A heritability of 0 implies that no genetic effects influence the observed variation in the trait, while a heritability of 1 implies that all of the variation in the trait is explained by the genetic effects. In general, the amount of information provided by the phenotype about the genetic risk is determined by the heritability. Note that heritability is population-specific and a heritability of 0 does not necessarily imply that there is no genetic determinism for the trait. 
<br />
<br />

## Brief Introduction to the `qgg` R package

The practical is based on the R package `qgg` (Rohde et al. (2021, 2022)). This package provides an infrastructure for efficient processing of large-scale genetic and phenotypic data including core functions for: 

* fitting linear mixed models 
* constructing genetic relationship matrices 
* estimating genetic parameters (heritability and correlation) 
* performing genomic prediction and genetic risk profiling 
* single or multi-marker association analyses

`qgg` handles large-scale data by taking advantage of:

* multi-core processing using openMP
* multithreaded matrix operations implemented in BLAS libraries (e.g., OpenBLAS, ATLAS or MKL)
* fast and memory-efficient batch processing of genotype data stored in binary files (i.e., PLINK bedfiles)

You can install qgg from CRAN with:

```{r,  eval=FALSE, echo=TRUE}
install.packages("qgg")
```
<br />

#### Input data/objects commonly used in the `qgg` package  {.unlisted .unnumbered}
All functions in `qgg` used for analysis of complex traits relies on a simple data infrastructure that takes the following main input: 

`y`:\qquad \quad vector, matrix or list of phenotypes\newline
`X`:\qquad \quad design matrix for non-genetic factors \newline
`W`:\qquad \quad matrix of centered and scaled genotypes (in memory) \newline
`Glist`:\quad list structure providing information on genotypes, sparse LD, and LD scores (on disk) \newline
`stat`:\quad data frame with marker summary statistics \newline
`sets`:\quad list of sets with marker ids \newline
`ids`:\quad vector of ids of individuals \newline
`rsids`:\quad vector marker marker ids \newline
<br />
<br />

## Session 1: Downloading the data using R

In this practical we will compute PGS based on simulated data. The data consist of disease phenotype, covariates, and SNP data. The data used in this practical are intended for demonstration purposes only. 
<br />

#### Load required packages: 
```{r,  eval=TRUE, echo=TRUE}
library(data.table)
```
<br />

#### Create (your own) directory for downloading files:
\vspace{-1truemm}
```{r,  eval=FALSE, echo=TRUE}
dir.create("./PGS")
```
<br />

#### Set (your own) working directory for the downloaded files:
\vspace{-1truemm}
```{r,  eval=F, echo=TRUE}
setwd("./PGS")
```
<br />

#### Download PLINK genotype files (bedfile, bimfile, famfile) from github repository: 
\vspace{-1truemm}
Genetic data are commonly stored in a binary format (as used by the software PLINK), named `.bed`-files. These files must be accompanied by `.bim` (contains information about the genetic variants) and `.fam` (contains information about the individuals) files. Read more about these file formats here:

-  https://www.cog-genomics.org/plink/1.9/formats#bed
-  https://www.cog-genomics.org/plink/1.9/formats#bim
-  https://www.cog-genomics.org/plink/1.9/formats#fam
<br />

```{r,  eval=FALSE, echo=TRUE}
url <- "https://github.com/PDRohde/AAU-human-genomics/raw/main/exercises/human.bed"
download.file( url=url, mode = "wb",  destfile="human.bed")
url <- "https://github.com/PDRohde/AAU-human-genomics/raw/main/exercises/human.bim"
download.file( url=url, destfile="human.bim")
url <- "https://github.com/PDRohde/AAU-human-genomics/raw/main/exercises/human.fam"
download.file( url=url, destfile="human.fam")
```
<br />
Note that `mode="wb"` for downloading the human.bed file. This is needed or otherwise the bed-file will be corrupted. If the data file is corrupted it can cause errors in the analyses. 
<br />

#### Download pheno and covar files from github repository;
\vspace{-1truemm}
```{r,  eval=FALSE, echo=TRUE}
url <- "https://github.com/PDRohde/AAU-human-genomics/raw/main/exercises/human.pheno"
download.file( url=url, destfile="human.pheno")
url <- "https://github.com/PDRohde/AAU-human-genomics/raw/main/exercises/human.covar"
download.file( url=url, destfile="human.covar")
```
<br />
<br />

## Session 2: Preparing the phenotype and covariable data using R

One of the first thing to do is to prepare the phenotypic data used in the analysis. The goal is to understand the variables, how many records the data set contains, how many missing values, what is the variable structure, what are the variable relationships and more. 

Several functions can be used (e.g., `str()`, `head()`, `dim()`, `table()`,`is.na()`). 
```{r,  eval=TRUE, echo=TRUE}
library(data.table)
```
<br />

#### Read phenotype and covariables data files
\vspace{-1truemm}
```{r,  eval=TRUE, echo=TRUE}
pheno <- fread(input="human.pheno", 
               data.table=FALSE)
```
<br />

```{r,  eval=TRUE, echo=TRUE}
covar <- fread(input="human.covar", 
               data.table=FALSE)
```
<br />

#### How many observations and which variables do we have in the data set? 
To get an overview of the data set you are working with you can use the `str()` or `head()` functions:
```{r,  eval=TRUE, echo=TRUE}
str(pheno)
str(covar)
```
<br />

```{r,  eval=TRUE, echo=TRUE}
head(pheno)
head(covar)
```
<br />

#### How is the phenotype distributed? 
Define the response variable
```{r,  eval=TRUE, echo=TRUE}
y <- pheno[,3]
names(y) <- pheno[,1]
```
<br />

Use the histogram and boxplot functions to visualize the distribution of the trait/covariables:

```{r,echo=T, fig.height=4,fig.width=8,fig.align='center'}
layout(matrix(c(1,2),ncol=2),widths=c(4,4),heights=4)
hist(y,las=1,cex.axis=.8,col="lightblue", ylab="Number of observations")
boxplot(covar[,4]~y, las=1,cex.axis=.8,col="lightblue",ylab="Values for covariate no 4")
```
<br />

#### Which factors or covariated influence the phenotype? 
The exploratory data analysis is the process of analyzing and visualizing the data to get a better understanding of the data. It is not a formal statistical test. Which factors should we include in the statistical model? To best answer these question we can fit a logistic regression model that include these factors in the model. 

This can be done using the `glm()` function:
  
```{r,  eval=TRUE, echo=TRUE}
fit <- glm( y ~ V3+V4+V5+V6+V7+V8+V9+V10+V11+V12+V13+V14, 
            data=covar, family=binomial(link="logit"))
summary(fit)
```
<br />
The exploration (including quality control) of phenotypes and covariables is a key step in quantitative genetic analyses. It is, however, beyond the scope of this practical. 
<br />
<br />


## Session 3: Prepare genotype for simulated data

The preparation (including quality control) of genotype data is a key step in quantitative genetic analyses. 

```{r,  eval=TRUE, echo=TRUE}
library(qgg)
```
<br />

#### Summarize genotype information in PLINK files {.unlisted .unnumbered} 
The function `gprep()` reads genotype information from binary PLINK files, and creates the `Glist` object that contains general information about the genotypes: 

```{r,  eval=T, echo=TRUE}
bedfiles <- "human.bed"
bimfiles <- "human.bim"
famfiles <- "human.fam"

Glist <- gprep(study="Example",
               bedfiles=bedfiles,
               bimfiles=bimfiles,
               famfiles=famfiles)
saveRDS(Glist, file="Glist.RDS", compress=FALSE)
```
```{r}
Glist <- readRDS("Glist.RDS")
```
<br />

The output from `gprep()` (`Glist`) has a list structure that contains information about the genotypes in the binary file. `Glist` is required for downstream analyses provided in the qgg package. Typically, the `Glist` is prepared once, and saved as an *.RDS-file. To explore the content of the `Glist` object:

```{r,  eval=TRUE, echo=TRUE}
names(Glist)
str(Glist)
```
<br />

#### Quality control of genotype data
In general it advisable to perform quality control of the genotype data. The quality control include removing markers with low genotyping rate, low minor allele frequency, not in Hardy-Weinberg Equilibrium. The function `gfilter()` can be used for filtering of markers:
```{r,  eval=TRUE, echo=TRUE}
rsids <-  gfilter( Glist = Glist,
                   excludeMAF=0.05,
                   excludeMISS=0.05,
                   excludeCGAT=TRUE,
                   excludeINDEL=TRUE,
                   excludeDUPS=TRUE,
                   excludeHWE=1e-12,
                   excludeMHC=FALSE)
```
<br />
The `gfilter()` function output the number of variants removed in the different quality control steps.
<br />
<br />

## Session 4: Compute GWAS summary statistics
One of the first step in PGS analyses is to generate or obtain GWAS summary statistics. Ideally these will correspond to the most powerful GWAS results available on the phenotype under study. In this example, we will use GWAS on the simulated disease phenotype. We will use only a subset of the data (training data) in the GWAS and the remaining subset of the data (validation data) to assess the accuracy of the polygenic scores. In the example below we only compute summary statistics for the markers that fulfil the quality control criteria.   

#### Define the response variable
```{r,  eval=TRUE, echo=TRUE}
y <- pheno[,3]
names(y) <- pheno[,1]
```
<br />

#### Create design matrix for the explanatory variables
```{r,  eval=TRUE, echo=TRUE}
X <- model.matrix(~V3+V4+V5+V6+V7+V8+V9+V10+V11+V12+V13+V14, data=covar)
rownames(X) <- covar$V1
X <- X[names(y),]
sum(names(y)%in%rownames(X))
```
<br />

#### Define training and validation samples
```{r,  eval=TRUE, echo=TRUE}
train <- sample(names(y),4000)
valid <- names(y)[!names(y)%in%train]
```
<br />

#### Computation of GWAS summary statistics 
The function `glma` can be used for computing GWAS summary statistics. Currently this function only fit a simple linear regression model, but we plan to add further modeling approached in a future release. 
```{r,  eval=TRUE, echo=TRUE}
stat <- glma(y=y[train], X=X[train,], Glist=Glist)
```
<br />

#### Explore the output (stat) form the `glma` function:
```{r,  eval=TRUE, echo=TRUE}
dim(stat)
head(stat)
```
<br />
<br />

## Session 5: Compute sparse LD matrices

Polygenic scoring based on summary statistics require the construction of a reference linkage disequilibrium (LD) correlation matrix. The LD matrix corresponds to the correlation between the genotypes of genetic variants across the genome. Here we use a sparse LD matrix approach using a fixed window approach (e.g. number of markers, 1 cM or 1000kb), which sets LD correlation values outside this window to zero. 

The function `gprep` can be used to compute sparse LD matrices which are stored on disk. The $r^2$ metric used is the pairwise correlation between markers (allele count alternative allele) in a specified region of the genome. Although this step can be slow unless R is linked to a fast BLAS it is typically only done once (or a few times). 

#### Define filenames for the sparse LD matrices. 
```{r,  eval=FALSE, echo=TRUE}
ldfiles <- "human.ld"
```
<br />

#### Compute sparse LD using only the filtered rsids (please see below first!)
```{r,  eval=FALSE, echo=TRUE}
Glist <- gprep( Glist,
                task="sparseld",
                msize=1000,
                rsids=rsids,
                ldfiles=ldfiles,
                overwrite=TRUE)
saveRDS(Glist, file="Glist_sparseLD_1k.RDS", compress=FALSE)
```
<br />


Note, it may take long time to compute LD, therefore, you can download the ld-file and updated Glist below. Remember to read the new `Glist`.
```{r,  eval=FALSE, echo=TRUE}
url <- "https://www.dropbox.com/scl/fi/5g9i631gq67wunk6njptr/human.ld?rlkey=00yk4oev7drk33syp6cjjke8j&dl=0"
download.file( url=url, destfile="human.ld")
url <- "https://www.dropbox.com/scl/fi/r23vhgt5okw37gohz71vu/Glist_sparseLD_1k.RDS?rlkey=ktv1p427kct9j13sydnn5k2a3&dl=0"
download.file( url=url, destfile="Glist_sparseLD_1k.RDS")
```
<br />
```{r, eval=T, echo=F}
Glist <- readRDS("Glist_sparseLD_1k.RDS")
```


<br />
<br />

## Session 6: Compute PGS using clumping and thresholding (C+T) 

Polygenic scoring using clumping and thresholding is a relative simple and robust method. Linkage disequilibrium makes identifying the contribution from causal independent genetic variants extremely challenging. One way of approximately capturing the right level of causal signal is to perform clumping, which removes markers in ways that only weakly correlated SNPs are retained but preferentially retaining the SNPs most associated with the phenotype under study. The clumping procedure uses a statistic (usually $P$-value) to sort the markers by importance (e.g. keeping the most significant ones). It takes the first one (e.g. most significant marker) and removes markers (i.e. set their effect to zero) if they are too correlated (e.g. $r^2>0.9$) with this one in a window around it. As opposed to pruning, this procedure makes sure that this marker is never removed, keeping at least one representative marker by region of the genome. Then it goes on with the next most significant marker that has not been removed yet. 

#### Clumping and thresholding
Clumping can be performed using the `adjStat()`-function in `qgg`. The input to the function is the summary statistic (`stat`), information about sparse LD matrices which is in the `Glist`, a threshold of linkage disequilibrium (e.g. $r^2=0.9$) and thresholds for $P$-values (`threshold = c(0.001, 0.05, ...)`):
```{r,  eval=TRUE, echo=TRUE}
threshold <- c(0.00001, 0.0001, 0.001, 0.005, 0.01, 0.05, 0.1, 0.2, 0.5,1)
statAdj <- adjStat(Glist=Glist, stat=stat, r2=0.9, threshold=threshold)
```
<br />

Explore the output (statAdj) using the `head` function:
```{r,  eval=TRUE, echo=TRUE}
head(statAdj)
```
<br />

A plot of the un-adjusted marker effect (from the `stat` data frame) against the adjusted marker effects (from the the `statAdj` data frame) illustrates that the C+T procedure keep only the most significant marker effects and is setting a large number of marker effects to zero (i.e. remove their effect).

```{r,  eval=T, echo=T,fig.height=4,fig.width=4,fig.align='center'}
plot( y=statAdj[rownames(stat),"b_0.001"], bg="lightblue",pch=21,las=1, cex.axis=.8,
     x=stat$b,
     xlab="Marginal Effect",
     ylab="Adjusted Effect",
     frame.plot=FALSE, ylim=c(-0.05,0.05), xlim=c(-0.05,0.05),
     main="Shrinkage using C+T \n (p=0.001, r2=0.9)")
```
<br />

#### Compute polygenic scores
For each of the `P`-value thresholds chosen in the C+T procedure a PGS is computed as:
			$$PGS=\sum_{i=1}^mX_i  \hat{b}_i$$
where $X_i$ is the genotype vector, and $\hat{b}_i$ the weight of the i’th single genetic marker.
The PGS are computed using the `gscore()` function. The input to the function is the adjusted summary statistic (`adjStat`), and information about the genotypes which are in the `Glist`:
```{r,  eval=eval, echo=TRUE}
pgs <- gscore(Glist=Glist,stat=statAdj)
```
<br />

#### Explore polygenic scores
It is always important to explore the PGS computed.  
```{r,  eval=eval, echo=TRUE,fig.height=7,fig.width=7,fig.align='center'}
head(pgs)
cor(pgs)
layout(matrix(1:4,ncol=2, byrow=TRUE))
hist(pgs[,"b"], main="PGS raw", las=1, col="lightblue", xlab="PGS - all individuals")
hist(pgs[,"b_0.001"], main="PGS P: 0.0001", las=1, col="lightblue", xlab="PGS - all individuals")
hist(pgs[valid,"b"], main="PGS raw", las=1, col="lightblue", xlab="PGS - validation set")
hist(pgs[valid,"b_0.001"], main="PGS P: 0.0001", las=1, col="lightblue", xlab="PGS - validatoin set")
```
<br />

#### Evalute polygenic scores 
The $P$-value threshold that provides the "best-fit" PGS under the C+T method is usually unknown. To approximate the "best-fit" PGS, we can perform a regression between PGS calculated at a range of $P$-value thresholds and then select the PGS that explains the highest proportion of phenotypic variance (e.g. R2) or has the highest AUC. This can be achieved using `acc()`-function as follows:

```{r,  eval=TRUE, echo=TRUE}
paCT <- acc(yobs=y[valid], ypred=pgs[valid,], typeoftrait="binary")
paCT
```
<br />

```{r,  eval=eval, echo=TRUE,fig.height=4,fig.width=5,fig.align='center'}
par(xpd=T)
bp <- barplot(paCT[-1,"Corr"], ylab="Correlation (y,PGS)", xlab="P-value cutoff", las=1, cex.axis=.8,col=rev(colorRampPalette(c("#FFFFCCFF", "#FF2A00FF"))(nrow(paCT)-1)),names=NA)
text(x=bp,srt=45,adj=1,y=-.005,labels=gsub("b_","",rownames(paCT)[-1]),cex=.8)
```
<br />

In comparison, we can see how strongly correlated the PGS is with $y$ for those individuals we utilised within the GWAS.
```{r,  eval=TRUE, echo=TRUE}
paCT <- acc(yobs=y[train], ypred=pgs[train,], typeoftrait="binary")
```
<br />

```{r,  eval=eval, echo=TRUE,fig.height=4,fig.width=5,fig.align='center'}
par(xpd=T)
bp <- barplot(paCT[-1,"Corr"], ylab="Correlation (y,PGS)", xlab="P-value cutoff", las=1, cex.axis=.8,col=rev(colorRampPalette(c("#FFFFCCFF", "#FF2A00FF"))(nrow(paCT)-1)),names=NA)
text(x=bp,srt=45,adj=1,y=-.05,labels=gsub("b_","",rownames(paCT)[-1]),cex=.8)
```
<br />

#### Plot polygenic scores 
For visualization, the PGS can be divided into groups (e.g., deciles), and the disease prevalence within each group was computed.

```{r,  eval=eval, echo=TRUE,fig.height=4,fig.width=7,fig.align='center'}
yobs <- y[valid]
ypred <- pgs[names(y[valid]),which.max(paCT[,"AUC"])]

nbin <- 10
qsets <- qgg:::splitWithOverlap( names(ypred)[order(ypred)],length(ypred)/nbin,0)
qy <- sapply(qsets,function(x){mean(yobs[x])})
qg <- sapply(qsets,function(x){mean(ypred[x])})

colfunc <- colorRampPalette(c("lightblue", "darkblue"))

layout(matrix(c(1,2),ncol=2),heights=c(4),widths=c(4,4))
plot(y=qy,x=qg,pch=19,ylab="Proportion of cases",xlab="Mean PGS", col=colfunc(nbin), frame.plot=FALSE, las=1,cex.axis=.8)
plot(y=qy,x=(1:nbin)/nbin,pch=19,ylab="Proportion of cases",xlab="Percentile of PGS", col=colfunc(nbin), frame.plot=FALSE, las=1,cex.axis=.8)
```
